### PR TITLE
feat(render_assets): optional support for resvg

### PR DESCRIPTION
**Pull request by actionless**

### CODE REVIEW

**Comment by nana-4** [comment]
As an overall script issue, these scripts don't work at all when `rendersvg` doesn't exist (but `inkscape` and `optipng` exist).

File: src/gtk/render-asset.sh [hunk 1]
```diff
@@ -1,23 +1,35 @@
...
 echo "Rendering '$ASSETS_DIR/$i@2.png'"
-"$INKSCAPE" --export-id="$i" \
-            --export-id-only \
-            --export-dpi=192 \
-            --export-png="$ASSETS_DIR/$i@2.png" "$SRC_FILE" >/dev/null \
-&& "$OPTIPNG" -o7 --quiet "$ASSETS_DIR/$i@2.png"
+if [[ -n "${RENDER_SVG}" ]] ; then
+  "$RENDER_SVG" --export-id "$i" \
+        --dpi 192 \
+         "$SRC_FILE" "$ASSETS_DIR/$i.png"
```
Comment (nana-4) [inline src/gtk/render-asset.sh]: `@2` is missing after `$i`.

File: src/_sass/_colors.scss.template [hunk 2]
```diff
@@ -62,6 +63,7 @@ $checked_overlay_color:     gtkalpha(currentColor, 0.24);
 $primary_color:            %SEL_BG%;
 $alt_primary_color:        %SEL_BG2%;
 $selected_overlay_color:    rgba($primary_color, 0.32);
+$selected_checkradio_color:  %ACCENT_BG%;
```
Comment (nana-4) [inline src/_sass/_colors.scss.template]: Why is this needed?

**Comment by actionless** [comment]
in arch linux test with `resvg-git`, it's newer than 0.4

File: src/_sass/_colors.scss.template (hunk 2 above)
Comment (actionless) [reply src/_sass/_colors.scss.template]: see commit message :-)

this stopped working after some recent Materia changes

Comment (actionless) [reply src/_sass/_colors.scss.template]: because the use of `-gtk-recolor()` it stopped working, to be precise

**Comment by actionless** [comment]
> As an overall script issue, these scripts don't work at all when rendersvg doesn't exist (but inkscape and optipng exist).

this i also fixed

File: src/_sass/_colors.scss.template (hunk 2 above)
Comment (nana-4) [reply src/_sass/_colors.scss.template]: Ah, sorry. I was going to tell it to you, but I'm going to drop the (teal) accent/secondary color on the next release (in order to follow the Material Design guidelines more accurately). So the blue `check` and `radio` are intentional.

Would you mind removing `ACCENT_BG` and the option in Oomox eventually? IIRC, that is only used for Materia-based theme. I think specifying only `SEL_BG` as an accent color is also good thing for compatibility with other oomox themes.

Comment (actionless) [reply src/_sass/_colors.scss.template]: it's used not only in Materia theme, but it's possible to remove it only for Materia theme[1] and make it use some other color instead[2]

[1] https://github.com/themix-project/oomox/blob/master/plugins/theme_materia/oomox_plugin.py#L44
[2] https://github.com/themix-project/oomox/blob/master/plugins/theme_materia/oomox_plugin.py#L80-L87

**Comment by actionless** [comment]
if dpi not working yet -- i suggest adding a workaround with zoom so far before it will be fixed

the benefits of switching from inkscape are very high: not only generation time more than 5 times faster but also shipping Themix Materia Plugin in Flatpak will became possible

File: src/_sass/_colors.scss.template (hunk 2 above)
Comment (actionless) [reply src/_sass/_colors.scss.template]: so, should i remove Accent option from Materia plugin completely?

Comment (nana-4) [reply src/_sass/_colors.scss.template]: > it's used not only in Materia theme

Ah, right, sorry for my misunderstanding.

> so, should i remove Accent option from Materia plugin completely?

Hmm, yes, please. :pray: 

(Considering the Material Design guidelines, I think it's better. Also since changing the color of specific SVG assets/ID to `ACCENT_BG` is quite difficult now.)

**Comment by nana-4** [comment]
I tried `resvg-git` in Arch Linux and confirmed the problem with extra spacing has been fixed. And yes, the dpi problem has not been fixed yet.

> dpi i haven't tested yet

You can check it by comparing the sizes of `foo.png` and `foo@2.png` in [`./src/gtk/assets`](https://github.com/nana-4/materia-theme/tree/master/src/gtk/assets) :)

> if dpi not working yet -- i suggest adding a workaround with zoom so far before it will be fixed

+1 :+1: Could you implement that, please?

> the benefits of switching from inkscape are very high: not only generation time more than 5 times faster but also shipping Themix Materia Plugin in Flatpak will became possible

Indeed. Since the remaining rendering problems are relatively small, I'd be happy to merge this (after some additional fixes) :)

I'm gonna add some additional reviews I forgot to mention. Please also check it. (Sorry for the inconvenience.)

**Comment by actionless** [comment]
thanks for checking, i'll address them later today

File: src/_sass/_colors.scss.template [hunk 3]
```diff
@@ -39,7 +40,7 @@ $inverse_fill_color:                  rgba(%SEL_FG%, 0.04);
 $bg_color:             %BG%;
 $base_color:           %TXT_BG%;
 $alt_base_color:       %INACTIVE_TXT_BG%;
-$surface_color:        %TXT_BG%;
+$surface_color:        %BTN_BG%;
```
Comment (nana-4) [inline src/_sass/_colors.scss.template]: To be honest, I'm not sure if this is OK, because `$surface_color` is used not only for buttons, but also for menus, popovers, popup dialogs, etc. (Just FYI, I'm also considering consolidating `$base_color` and `$surface_color` eventually as per Material Design guidelines.)

What do you think about this? If you are OK, I would not oppose this change.

File: src/_sass/_colors.scss.template [hunk 4]
```diff
@@ -17,7 +17,8 @@ $disabled_fg_color:           rgba(%FG%, 0.5);
 $disabled_secondary_fg_color: rgba(%FG%, 0.3);
 $track_color:                 rgba(%FG%, 0.3);
 $divider_color:               rgba(%FG%, 0.12);
-$fill_color:                  rgba(%FG%, 0.04);
+//$fill_color:                  rgba(%FG%, 0.04);
+$fill_color:                  %TXT_BG%;
```
Comment (nana-4) [inline src/_sass/_colors.scss.template]: In Material Design, filled text fields have a translucent background color derived from text color.
(See https://material.io/design/components/text-fields.html for details.)

Can you undo the changes?

Comment (actionless) [reply src/_sass/_colors.scss.template]: i've read the article and i think your design decision could be conflicting with the following requirement: `Don’t design text fields to look similar to buttons, as they could be mistaken for buttons.`

and currently both button and inputs are using fill_color

i think the best option will be if you define in _colors.scss just names of the elements (button, input, checkbox, whatever) and next just set them as needed, let's say `$button_bg = $fill_color` and so on, that will allow to override any element without affecting others

Comment (nana-4) [reply src/_sass/_colors.scss.template]: > i've read the article and i think your design decision could be conflicting with the following requirement: Don’t design text fields to look similar to buttons, as they could be mistaken for buttons.

Personally I don't have the concern, because buttons never have the 1px grey underline + square bottom corners + fill. In the "**Don’t**" example in the article, underline can not be seen.

> and currently both button and inputs are using fill_color

Yes, but `$fill_color` is only used for the **disabled** raised button, not for other types of button.

> i think the best option will be if you define in _colors.scss just names of the elements (button, input, checkbox, whatever) and next just set them as needed, let's say $button_bg = $fill_color and so on, that will allow to override any element without affecting others

I also had that idea, but now I'm wondering if we really need to do that because [the baseline Material color theme](https://material.io/design/color/the-color-system.html#color-theme-creation) is composed only in a few colors. For example, there is such a description:

> Surface colors affect surfaces of components, such as cards, sheets, and menus.

Comment (actionless) [reply src/_sass/_colors.scss.template]: yeah, sorry for being unclear, i exactly meant disabled style (the issue is quite observable since disabled button is often used in UIs, for example in GtkFileChooser unless some file is chosen)

> but now I'm wondering if we really need to do that because the baseline Material color theme is composed only in a few colors. For example, there is such a description:
>
>    Surface colors affect surfaces of components, such as cards, sheets, and menus.


i think such change won't contradict to the guidelines, since the result will be same. It's more about abstraction level: 

1) currently what you have 
```scss
colors.scss:
$surface = #ff0000;

widgets' scss:
card {
  background: $surface;
}
sheet {
  background: $surface;
}
menu {
  background: $surface;
}
```

2) what i propose:

colors.scss:
$surface = #ff0000;
$card_bg = $surface;
$sheet_bg = $surface;
$menu_bg = $surface;

widgets' scss:
card {
  background: $card_bg;
}
sheet {
  background: $sheet_bg;
}
menu {
  background: $menu_bg;
}
```

apart from being recoloring-friendly having widgets colors defined explicitly will make it more clearly defined which widgets have which colors without guessing or `grep`-ing

Comment (nana-4) [reply src/_sass/_colors.scss.template]: > (the issue is quite observable since disabled button is often used in UIs, for example in GtkFileChooser unless some file is chosen)

Hm... I'd say people never mistake a disabled button for an entry, because disabled button doesn't have the underline and square bottom corners, which entry has.

Even if this is still an issue, I'd say this is rather an upstream Material Design issue, because entry and disabled button have such styles in that design.

> i think such change won't contradict to the guidelines, since the result will be same. It's more about abstraction level:
apart from being recoloring-friendly having widgets colors defined explicitly will make it more clearly defined which widgets have which colors without guessing or grep-ing

I can understand it, but in reality, it requires a larger amount of work, for example:


```scss
widgets' scss:
card {
  background: $card_bg;
  color: $card_fg;
  :disabled { color: $card_disabled_fg; }
  .hint-text { color: $card_secondary_fg; }
  .track { background: $card_track; }
  .divider { background: $card_divider; }
  entry, button:disabled { background: $card_fill; }
}
sheet {
  background: $sheet_bg;
  color: $sheet_fg;
  :disabled { color: $sheet_disabled_fg; }
  .hint-text { color: $sheet_secondary_fg; }
  .track { background: $sheet_track; }
  .divider { background: $sheet_divider; }
  entry, button:disabled { background: $sheet_fill; }
}
menu {
  background: $menu_bg;
  color: $menu_fg;
  :disabled { color: $menu_disabled_fg; }
  .hint-text { color: $menu_secondary_fg; }
  .track { background: $menu_track; }
  .divider { background: $menu_divider; }
  entry, button:disabled { background: $menu_fill; }
}
```

Especially, gnome-shell CSS and cinnamon CSS are quite complicated, so that refactoring is never easy for me.

Also I'm wondering even if oomox users really want to recolor widgets in detail **with Materia**. As far as I know, most of the people who customize Materia change only basic colors such as accent color, titlebar color, and HSV of whole background colors. (For instance, a person who want to change only the button to "dark" and leave the rest "light" is pretty rare with Materia-based theme, I guess.) Since people who use Materia-based theme are supposed to be more or less fans of Material Design, can't we bind them to the guidelines to some extent?

Comment (nana-4) [reply src/_sass/_colors.scss.template]: To clarify, I share my idea of the _ideal_ Oomox recolor options for Materia:

- `Background`
- `Text`
- `Surface (e.g. Buttons, Menus, Popovers)`
- `View`
- `Titlebar Background`
- `Titlebar Text`
- ? `Error` (This might be unnecessary.)
- `Accent`
- `Selection Opacity` [default value: `0.32`]
- `Translucent Panel Opacity` [default value: `0.6`]

Comment (actionless) [reply src/_sass/_colors.scss.template]: i think the problem above could be fixed with mixins and/or functions (for example in numix/oomox border is generated by function), but i am agree it could be too big volume of refactoring need to be done from your side to implement those

so better let's stick to the list of options which you are proposing in the next message

Comment (actionless) [reply src/_sass/_colors.scss.template]: also side note, for Error, Warning, Info themix gui using Red, Yellow and Blue terminal colors accordingly, mb to make it clear i will rename "Red" into "Red (error)" and so on, what do you think?

Comment (actionless) [reply src/_sass/_colors.scss.template]: btw i did reverted all scss changes

and with the current _colors.scss buttons receiving TXT_BG color, not BTN_BG

Comment (actionless) [reply src/_sass/_colors.scss.template]: reverted one of revert commits :) please let me know if it's fine now

Comment (nana-4) [reply src/_sass/_colors.scss.template]: > reverted one of revert commits :) please let me know if it's fine now

LGTM! Thanks a lot!

> also side note, for Error, Warning, Info themix gui using Red, Yellow and Blue terminal colors accordingly, mb to make it clear i will rename "Red" into "Red (error)" and so on, what do you think?

As a temporary fix, I think it's fine. Or maybe "Red (error color for GTK+)" sounds more clear (because users usually don't know the error color option is there).

However, what do you think about separating it into Theme "Error" option and Terminal "Red" option, and so on? Because:

- In many cases, these options are hidden in the GUI. 
- Users are hard to notice that the "Theme" color options are in the "Terminal" section.
- Some users may not want "Theme Error color" and "Terminal Red color" to have the same color.

As some themix variables have default values, applying the theme's default error/warning/info colors to the default value of those variables may be good for users and for ease of implementation.

Comment (actionless) [reply src/_sass/_colors.scss.template]: > In many cases, these options are hidden in the GUI.
> Users are hard to notice that the "Theme" color options are in the "Terminal" section.

those i think could be solved by duplicating the values so it will be presented twice, in theme and terminal sections. however i am not sure those options are so common to be fine-tuned to extend already big list of theme options (btw in GUI suggestions topic it would be good to discuss how better to split that list into some tabs or expanders to make it easier to navigate)

> Some users may not want "Theme Error color" and "Terminal Red color" to have the same color.

that easily could be managed by generating different themes for GTK and terminal (since their output is not connected anyway)

Comment (actionless) [reply src/_sass/_colors.scss.template]: also tonight i will change the list of available options in Materia Themix plugin as you suggested in https://github.com/nana-4/materia-theme/pull/329#discussion_r242819290, after that i suggest you playing with it a bit to see if theme preview and generation are happening as you expect them to happen

Comment (nana-4) [reply src/_sass/_colors.scss.template]: > however i am not sure those options are so common to be fine-tuned to extend already big list of theme options (btw in GUI suggestions topic it would be good to discuss how better to split that list into some tabs or expanders to make it easier to navigate)

Indeed. So let's discuss about it in that topic :)

> also tonight i will change the list of available options in Materia Themix plugin as you suggested in #329 (comment), after that i suggest you playing with it a bit to see if theme preview and generation are happening as you expect them to happen

Thanks for accepting my proposal :) Yes, I'll try it.

Comment (actionless) [reply src/_sass/_colors.scss.template]: done, gonna clean-up the code a bit and push in few minutes:

2018-12-19--1545254083_1342x816_scrot

Comment (actionless) [reply src/_sass/_colors.scss.template]: you can test now with `oomox-git` AUR package (however some of Materia-related stuff is not merged yet, so those two options won't be applied in generated theme til then):

https://github.com/nana-4/materia-theme/pull/331/files

Comment (nana-4) [reply src/_sass/_colors.scss.template]: Great job! BTW, are you planning to change the descriptions and reorder the options as https://github.com/nana-4/materia-theme/pull/329#discussion_r242819290 on the Materia theme style?

For example, in Materia, the description "Menu/toolbar background" is incorrect. That option only changes `headerbar`/`titlebar`/`menubar`, never changes `menu`/`toolbar`.

Comment (nana-4) [reply src/_sass/_colors.scss.template]: Also in my opinion, "Selection Opacity" and "Translucent Panel Opacity" are rather "Color options", not "Other option". Is it possible to reorder them to the bottom of the "color section"?

Comment (actionless) [reply src/_sass/_colors.scss.template]: so far reordering not possible because the theme keys used there are also used in other themes

solution could be to define new theme vars for those materia-specific options, like View or Surface

and in config to used TXT_BG as fallback value for VIEW_BG and so on
